___

<p style="text-align: center;"><img src="https://docs.google.com/uc?id=1lY0Uj5R04yMY3-ZppPWxqCr5pvBLYPnV" class="img-fluid" 
alt="CLRSWY"></p>

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#9d4f8c; font-size:100%; text-align:center; border-radius:10px 10px;">WAY TO REINVENT YOURSELF</p>

<img src=https://i.ibb.co/wJW61Y2/Used-cars.jpg width="700" height="200">

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#060108; font-size:200%; text-align:center; border-radius:10px 10px;">The Capstone Project of Data Analytics Module</p>

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#060108; font-size:150%; text-align:center; border-radius:10px 10px;">Car Price Prediction EDA</p>

<a id="0"></a>
<p style="background-color:plum; color:floralwhite; font-size:175%; text-align:center; border-radius:10px 10px; font-family:newtimeroman; line-height: 1.4; font-weight:bold; text-transform:capitalize;">Introduction</p>

<a href="#content" class="btn btn-primary btn-sm" role="button" aria-pressed="true" style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

<p style="font-size:22px; font-weight:bold;">Introduction</p> 
Welcome to "***AutoScout Data Analysis Project***". This is the capstone project of ***Data Analysis*** Module. **Auto Scout** data which using for this project, scraped from the on-line car trading company in 2019, contains many features of 9 different car models. In this project, you will have the opportunity to apply many commonly used algorithms for Data Cleaning and Exploratory Data Analysis by using many Python libraries such as Numpy, Pandas, Matplotlib, Seaborn, Scipy you will analyze clean dataset.

### Some Reminders on Exploratory data analysis (EDA)

Exploratory data analysis (EDA) is an especially important activity in the routine of a data analyst or scientist. It enables an in depth understanding of the dataset, define or discard hypotheses and create predictive models on a solid basis. It uses data manipulation techniques and several statistical tools to describe and understand the relationship between variables and how these can impact business. By means of EDA, we can obtain meaningful insights that can impact analysis under the following questions (If a checklist is good enough for pilots to use every flight, it’s good enough for data scientists to use with every dataset).
1. What question are you trying to solve (or prove wrong)?
2. What kind of data do you have?
3. What’s missing from the data?
4. Where are the outliers?
5. How can you add, change or remove features to get more out of your data?

**``Exploratory data analysis (EDA)``** is often an **iterative brainstorming process** where you pose a question, review the data, and develop further questions to investigate before beginning model development work. The image below shows how the brainstorming phase is connected with that of understanding the variables and how this in turn is connected again with the brainstorming phase.<br>

<img src=https://i.ibb.co/k0MC950/EDA-Process.png width="300" height="100">

[Image Credit: Andrew D.](https://towardsdatascience.com/exploratory-data-analysis-in-python-a-step-by-step-process-d0dfa6bf94ee)

**``In this context, the project consists of 3 parts in general:``**
* **The first part** is related to 'Data Cleaning'. It deals with Incorrect Headers, Incorrect Format, Anomalies, and Dropping useless columns.
* **The second part** is related to 'Filling Data', in other words 'Imputation'. It deals with Missing Values. Categorical to numeric transformation is done as well.
* **The third part** is related to 'Handling Outliers of Data' via Visualization libraries. So, some insights will be extracted.

**``NOTE:``**  However, you are free to create your own style. You do NOT have to stick to the steps above. We, the DA & DV instructors, recommend you study each part separately to create a source notebook for each part title for your further studies. 

<!-- # PART- 1 `( Data Cleaning )` -->

# PART- 1 Data Cleaning

<a id="2"></a>
<p style="background-color:plum; color:floralwhite; font-size:175%; text-align:center; border-radius:10px 10px; font-family:newtimeroman; line-height: 1.4; font-weight:bold; text-transform:capitalize;">Importing Related Libraries</p>

<a href="#content" class="btn btn-primary btn-sm" role="button" aria-pressed="true" style="color:blue; background-color:#dfa8e4" data-toggle="popover">Content</a>

- Once you've Installed NumPy and Pandas etc. you can Import them as a Library.
- Reading the Data from File

In [681]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib import patches
import seaborn as sns

import re 
import warnings
# Suppressing a warning 
warnings.filterwarnings("ignore") 
warnings.warn("this will not show")

# !pip install skimpy
from skimpy import clean_columns

In [682]:
# Suppressing a warning 
warnings.filterwarnings("ignore") 

# It is a magic function that renders the figure in the notebook
%matplotlib inline 
# %matplotlib notebook

plt.rcParams["figure.figsize"] = (10,6)
# plt.rcParams['figure.dpi'] = 100

plt.rcParams["figure.figsize"] = (10,6)
# plt.rcParams['figure.dpi'] = 100

# Changing the figure size of a seaborn axes 
sns.set(rc={"figure.figsize": (10, 6)})

# The style parameters control properties like the color of the background and whether a grid is enabled by default.
sns.set_style("whitegrid")

pd.set_option('display.float_format', lambda x: '%.3f' % x)

# To display maximum columns
pd.set_option('display.max_columns', None)

# To display maximum rows
pd.set_option('display.max_rows', 100)

# Introducing Data

##  Reading scout_car.json file as a dataframe named df

In [683]:
df =pd.read_json("../scout_car.json", lines=True)
df

url      make_model  \
0      https://www.autoscout24.com//offers/audi-a1-sp...         Audi A1   
1      https://www.autoscout24.com//offers/audi-a1-1-...         Audi A1   
2      https://www.autoscout24.com//offers/audi-a1-sp...         Audi A1   
3      https://www.autoscout24.com//offers/audi-a1-1-...         Audi A1   
4      https://www.autoscout24.com//offers/audi-a1-sp...         Audi A1   
...                                                  ...             ...   
15914  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15915  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15916  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15917  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15918  https://www.autoscout24.com//offers/renault-es...  Renault Espace   

                                       short_description body_type  price  \
0            Sportback 1.4 TDI S-tronic Xenon Navi Klima    Sedans  15770   
1                                         1.8 TFSI sport    Sedans  14500   
2      Sportback 1.6 TDI S tronic Einparkhilfe plus+m...    Sedans  14640   
3                                1.4 TDi Design S tronic    Sedans  14500   
4      Sportback 1.4 TDI S-Tronic S-Line Ext. admired...    Sedans  16790   
...                                                  ...       ...    ...   
15914              Blue dCi 200CV EDC Executive 4Control       Van  39950   
15915                TCe 225 EDC GPF LIM Deluxe Pano,RFK       Van  39885   
15916  Blue dCi 200 EDC Initiale Paris Leder LED Navi...       Van  39875   
15917  Blue dCi 200CV EDC Business , NUOVA DA IMMATRI...       Van  39700   
15918        INITIALE PARIS TCe 225 EDC GPF NAVI SHZ AUX       Van  40999   

                    vat         km registration         prev_owner  kW  \
0        VAT deductible  56,013 km      01/2016  2 previous owners NaN   
1      Price negotiable  80,000 km      03/2017               None NaN   
2        VAT deductible  83,450 km      02/2016   1 previous owner NaN   
3                  None  73,000 km      08/2016   1 previous owner NaN   
4                  None  16,200 km      05/2016   1 previous owner NaN   
...                 ...        ...          ...                ...  ..   
15914    VAT deductible       - km          -/-               None NaN   
15915    VAT deductible   9,900 km      01/2019   1 previous owner NaN   
15916    VAT deductible      15 km      03/2019   1 previous owner NaN   
15917    VAT deductible      10 km      06/2019               None NaN   
15918    VAT deductible       - km      01/2019   1 previous owner NaN   

           hp                                               Type  \
0       66 kW            [, Used, , Diesel (Particulate Filter)]   
1      141 kW                               [, Used, , Gasoline]   
2       85 kW            [, Used, , Diesel (Particulate Filter)]   
3       66 kW            [, Used, , Diesel (Particulate Filter)]   
4       66 kW            [, Used, , Diesel (Particulate Filter)]   
...       ...                                                ...   
15914  147 kW             [, New, , Diesel (Particulate Filter)]   
15915  165 kW  [, Used, , Super 95 / Super Plus 98 (Particula...   
15916  146 kW                       [, Pre-registered, , Diesel]   
15917  147 kW                       [, Pre-registered, , Diesel]   
15918  165 kW                      [, Demonstration, , Super 95]   

                                         Previous Owners  \
0                                                  \n2\n   
1                                                    NaN   
2                                                  \n1\n   
3                                                  \n1\n   
4                                                  \n1\n   
...                                                  ...   
15914                                                NaN   
15915  [\n1\n, \n, 7.4 l/100 

## Creating a copy from df named df_copy

In [684]:
df_copy = df.copy()

## Recognizing and Understanding Data

01. Reading the Data from File
02. Try to understand what the data looks like
03. Examining the Data
04. Fixing a data type
05. Combining object columns

## Try to understand what the data looks like
- Check the head, shape, data-types of the features.
- Check if there are some dublicate rows or not. If there are, then drop them. 
- Check the statistical values of features.
- Basically check the missing values. (NaN, None)
- If needed, rename the columns' names for easy use.
> Bu kısımda düzenleme yapacağım.

## Checking the head, shape, data-types of the features.

In [685]:
df.shape

(15919, 54)

## Checking the dataframe with head, tail and sample

In [686]:
df.head()

url make_model  \
0  https://www.autoscout24.com//offers/audi-a1-sp...    Audi A1   
1  https://www.autoscout24.com//offers/audi-a1-1-...    Audi A1   
2  https://www.autoscout24.com//offers/audi-a1-sp...    Audi A1   
3  https://www.autoscout24.com//offers/audi-a1-1-...    Audi A1   
4  https://www.autoscout24.com//offers/audi-a1-sp...    Audi A1   

                                   short_description body_type  price  \
0        Sportback 1.4 TDI S-tronic Xenon Navi Klima    Sedans  15770   
1                                     1.8 TFSI sport    Sedans  14500   
2  Sportback 1.6 TDI S tronic Einparkhilfe plus+m...    Sedans  14640   
3                            1.4 TDi Design S tronic    Sedans  14500   
4  Sportback 1.4 TDI S-Tronic S-Line Ext. admired...    Sedans  16790   

                vat         km registration         prev_owner  kW      hp  \
0    VAT deductible  56,013 km      01/2016  2 previous owners NaN   66 kW   
1  Price negotiable  80,000 km      03/2017               None NaN  141 kW   
2    VAT deductible  83,450 km      02/2016   1 previous owner NaN   85 kW   
3              None  73,000 km      08/2016   1 previous owner NaN   66 kW   
4              None  16,200 km      05/2016   1 previous owner NaN   66 kW   

                                      Type Previous Owners  \
0  [, Used, , Diesel (Particulate Filter)]           \n2\n   
1                     [, Used, , Gasoline]             NaN   
2  [, Used, , Diesel (Particulate Filter)]           \n1\n   
3  [, Used, , Diesel (Particulate Filter)]           \n1\n   
4  [, Used, , Diesel (Particulate Filter)]           \n1\n   

                         Next Inspection                      Inspection new  \
0  [\n06/2021\n, \n99 g CO2/km (comb)\n]               [\nYes\n, \nEuro 6\n]   
1                                    NaN                                 NaN   
2                                    NaN                                 NaN   
3                                    NaN                                 NaN   
4                                    NaN  [\nYes\n, \n109 g CO2/km (comb)\n]   

                           Warranty                      Full Service  \
0           [\n, \n, \n4 (Green)\n]                          [\n, \n]   
1                               NaN                               NaN   
2  [\n, \n, \n99 g CO2/km (comb)\n]                               NaN   
3                               NaN  [\n, \n, \n99 g CO2/km (comb)\n]   
4              [\n, \n, \nEuro 6\n]           [\n, \n, \n4 (Green)\n]   

    Non-smoking Vehicle null      Make         Model     Offer Number  \
0              [\n, \n]   []  \nAudi\n  [\n, A1, \n]  [\nLR-062483\n]   
1                   NaN   []  \nAudi\n  [\n, A1, \n]              NaN   
2                   NaN   []  \nAudi\n  [\n, A1, \n]   [\nAM-95365\n]   
3  [\n, \n, \nEuro 6\n]   []  \nAudi\n  [\n, A1, \n]              NaN   
4              [\n, \n]   []  \nAudi\n  [\n, A1, \n]      [\nC1626\n]   

  First Registration       Body Color      Paint Type  \
0     [\n, 2016, \n]  [\n, Black, \n]  [\nMetallic\n]   
1     [\n, 2017, \n]    [\n, Red, \n]             NaN   
2     [\n, 2016, \n]  [\n, Black, \n]  [\nMetallic\n]   
3     [\n, 2016, \n]  [\n, Brown, \n]  [\nMetallic\n]   
4     [\n, 2016, \n]  [\n, Black, \n]  [\nMetallic\n]   

            Body Color Original          Upholstery              Body  \
0           [\nMythosschwarz\n]  [\nCloth, Black\n]  [\n, Sedans, \n]   
1                           NaN   [\nCloth, Grey\n]  [\n, Sedans, \n]   
2  [\nmythosschwarz metallic\n]  [\nCloth, Black\n]  [\n, Sedans, \n]   
3                           NaN                 NaN  [\n, Sedans, \n]   
4  [\nMythosschwarz Metallic\n]  [\nCloth, Black\n]  [\n, Sedans, \n]   

  Nr. of Doors Nr. of Seats      Model Code         Gearing Type  \
0      [\n5\n]      [\n5\n]  [\n0588/BDF\n]  [\n, Automatic, \n]   
1      [\n3\n]      [\n4\n]  [\n0588/BCY\n]  [\n, Automatic, \n]   
2      [\n4\n]      

In [687]:
df.tail()

url      make_model  \
15914  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15915  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15916  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15917  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15918  https://www.autoscout24.com//offers/renault-es...  Renault Espace   

                                       short_description body_type  price  \
15914              Blue dCi 200CV EDC Executive 4Control       Van  39950   
15915                TCe 225 EDC GPF LIM Deluxe Pano,RFK       Van  39885   
15916  Blue dCi 200 EDC Initiale Paris Leder LED Navi...       Van  39875   
15917  Blue dCi 200CV EDC Business , NUOVA DA IMMATRI...       Van  39700   
15918        INITIALE PARIS TCe 225 EDC GPF NAVI SHZ AUX       Van  40999   

                  vat        km registration        prev_owner  kW      hp  \
15914  VAT deductible      - km          -/-              None NaN  147 kW   
15915  VAT deductible  9,900 km      01/2019  1 previous owner NaN  165 kW   
15916  VAT deductible     15 km      03/2019  1 previous owner NaN  146 kW   
15917  VAT deductible     10 km      06/2019              None NaN  147 kW   
15918  VAT deductible      - km      01/2019  1 previous owner NaN  165 kW   

                                                    Type  \
15914             [, New, , Diesel (Particulate Filter)]   
15915  [, Used, , Super 95 / Super Plus 98 (Particula...   
15916                       [, Pre-registered, , Diesel]   
15917                       [, Pre-registered, , Diesel]   
15918                      [, Demonstration, , Super 95]   

                                         Previous Owners  \
15914                                                NaN   
15915  [\n1\n, \n, 7.4 l/100 km (comb), \n, 9.2 l/100...   
15916                   [\n1\n, \n139 g CO2/km (comb)\n]   
15917                                                NaN   
15918                                              \n1\n   

                              Next Inspection  \
15914                                     NaN   
15915  [\n01/2022\n, \n168 g CO2/km (comb)\n]   
15916                                     NaN   
15917                                     NaN   
15918                                     NaN   

                           Inspection new                 Warranty  \
15914                                 NaN            \n24 months\n   
15915                                 NaN     [\n, \n, \nEuro 6\n]   
15916         [\nYes\n, \nEuro 6d-TEMP\n]  [\n, \n, \n4 (Green)\n]   
15917                                 NaN                       \n   
15918  [\nYes\n, \n153 g CO2/km (comb)\n]                      NaN   

      Full Service      Non-smoking Vehicle null         Make  \
15914          NaN                      NaN   []  \nRenault\n   
15915          NaN  [\n, \n, \n4 (Green)\n]   []  \nRenault\n   
15916     [\n, \n]                 [\n, \n]   []  \nRenault\n   
15917          NaN                      NaN   []  \nRenault\n   
15918          NaN                      NaN   []  \nRenault\n   

                  Model          Offer Number First Registration  \
15914  [\n, Espace, \n]        [\n10988301\n]                NaN   
15915  [\n, Espace, \n]     [\n507370_3223\n]     [\n, 2019, \n]   
15916  [\n, Espace, \n]        [\nEspace16\n]     [\n, 2019, \n]   
15917  [\n, Espace, \n]         [\n2691331\n]     [\n, 2019, \n]   
15918  [\n, Espace, \n]  [\nRe_30000008029\n]     [\n, 2019, \n]   

            Body Color      Paint Type           Body Color Original  \
15914   [\n, Grey, \n]  [\nMetallic\n]            [\nGrigio scuro\n]   
15915   [\n, Grey, \n]  [\nMetallic\n]     [\nStahl-Grau Metallic\n]   
15916  [\n, White, \n]             NaN             [\narktis-weiß\n]   
15917   [\n, Grey, \n]             NaN                  [\nGrigio\n]   
15918   [\n, Grey, \n]  [\nMetallic\n]  [\nTitanium-Grau Metallic\

In [688]:
df.sample(5)

url     make_model  \
11164  https://www.autoscout24.com//offers/opel-insig...  Opel Insignia   
2333   https://www.autoscout24.com//offers/audi-a1-sp...        Audi A1   
7851   https://www.autoscout24.com//offers/opel-astra...     Opel Astra   
4282   https://www.autoscout24.com//offers/audi-a3-sp...        Audi A3   
12696  https://www.autoscout24.com//offers/opel-insig...  Opel Insignia   

                                       short_description      body_type  \
11164                                    1.6 Turbo Start  Station wagon   
2333                            Sportback 25 TFSI S line        Compact   
7851   Sports Tourer Innovation MJ19 1.6 CDTI AT6 136...  Station wagon   
4282                               Sportback Dsl 1.6 TDi         Sedans   
12696        Grand Sport 1.6 Turbo Business Executive AT         Sedans   

       price               vat         km registration prev_owner  kW      hp  \
11164  10500  Price negotiable  60,000 km      08/2017       None NaN  125 kW   
2333   19497              None       - km          -/-       None NaN   70 kW   
7851   23465    VAT deductible       1 km          -/-       None NaN  100 kW   
4282   17899    VAT deductible  29,500 km      06/2017       None NaN   81 kW   
12696  26095    VAT deductible       - km          -/-       None NaN  147 kW   

                                         Type Previous Owners Next Inspection  \
11164                    [, Used, , Gasoline]             NaN             NaN   
2333                      [, New, , Gasoline]             NaN             NaN   
7851   [, New, , Diesel (Particulate Filter)]             NaN             NaN   
4282                       [, Used, , Diesel]             NaN             NaN   
12696                 [, New, , Super E10 95]             NaN             NaN   

                           Inspection new                           Warranty  \
11164                                 NaN                                NaN   
2333                                  NaN                                NaN   
7851                                  NaN  [\n, \n, \n126 g CO2/km (comb)\n]   
4282                                  NaN                      \n12 months\n   
12696  [\nYes\n, \n146 g CO2/km (comb)\n]                                NaN   

      Full Service         Non-smoking Vehicle null      Make  \
11164          NaN                         NaN   []  \nOpel\n   
2333           NaN                         NaN   []  \nAudi\n   
7851           NaN                         NaN   []  \nOpel\n   
4282           NaN                         NaN   []  \nAudi\n   
12696          NaN  [\n, \n, \nEuro 6d-TEMP\n]   []  \nOpel\n   

                    Model                                    Offer Number  \
11164  [\n, Insignia, \n]                                             NaN   
2333         [\n, A1, \n]                                   [\n2587866\n]   
7851      [\n, Astra, \n]                                     [\n12082\n]   
4282         [\n, A3, \n]  [\n134_46987e8c-09d7-4355-a63a-a8311dbfbd95\n]   
12696  [\n, Insignia, \n]                                     [\n71269\n]   

      First Registration       Body Color      Paint Type  \
11164     [\n, 2017, \n]   [\n, Grey, \n]  [\nMetallic\n]   
2333                 NaN              NaN             NaN   
7851                 NaN              NaN             NaN   
4282      [\n, 2017, \n]  [\n, Black, \n]             NaN   
12696                NaN              NaN             NaN   

                     Body Color Original          Upholstery  \
11164                                NaN                 NaN   
2333                                 NaN                 NaN   
7851                       [\nwählbar\n]                 NaN   
4282                    [\nZwart/Noir\n]                 NaN   
12696  [\nwählbar - ggf. mit Aufpreis\n]  [\nCloth, Other\n]   

                          Body Nr. of Doors Nr. of Seats      Model Code  \
11164  [\n, St

## Checking the summary information of df

In [689]:
# Checking the summary information of df with df.info()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 54 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   url                            15919 non-null  object 
 1   make_model                     15919 non-null  object 
 2   short_description              15873 non-null  object 
 3   body_type                      15859 non-null  object 
 4   price                          15919 non-null  int64  
 5   vat                            11406 non-null  object 
 6   km                             15919 non-null  object 
 7   registration                   15919 non-null  object 
 8   prev_owner                     9091 non-null   object 
 9   kW                             0 non-null      float64
 10  hp                             15919 non-null  object 
 11  Type                           15917 non-null  object 
 12  Previous Owners                9279 non-null  

 ## Checking the null values of df

In [690]:
df.isnull().sum()

url                                  0
make_model                           0
short_description                   46
body_type                           60
price                                0
vat                               4513
km                                   0
registration                         0
prev_owner                        6828
kW                               15919
hp                                   0
Type                                 2
Previous Owners                   6640
Next Inspection                  12384
Inspection new                   11987
Warranty                          5420
Full Service                      7704
Non-smoking Vehicle               8742
null                                 0
Make                                 0
Model                                0
Offer Number                      3175
First Registration                1597
Body Color                         597
Paint Type                        5772
Body Color Original      

### Checking if there are some dublicate rows or not. If there are, then drop them.

In [691]:
# error rising because of list type
# df.duplicated().sum()

In [692]:
df.duplicated
# As it is seen below, there is not any duplicated values.

<bound method DataFrame.duplicated of                                                      url      make_model  \
0      https://www.autoscout24.com//offers/audi-a1-sp...         Audi A1   
1      https://www.autoscout24.com//offers/audi-a1-1-...         Audi A1   
2      https://www.autoscout24.com//offers/audi-a1-sp...         Audi A1   
3      https://www.autoscout24.com//offers/audi-a1-1-...         Audi A1   
4      https://www.autoscout24.com//offers/audi-a1-sp...         Audi A1   
...                                                  ...             ...   
15914  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15915  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15916  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15917  https://www.autoscout24.com//offers/renault-es...  Renault Espace   
15918  https://www.autoscout24.com//offers/renault-es...  Renault Espace   

                                       short_desc

### Checking and cleaning the columns of dataframe

In [693]:
df.columns

Index(['url', 'make_model', 'short_description', 'body_type', 'price', 'vat',
       'km', 'registration', 'prev_owner', 'kW', 'hp', 'Type',
       'Previous Owners', 'Next Inspection', 'Inspection new', 'Warranty',
       'Full Service', 'Non-smoking Vehicle', 'null', 'Make', 'Model',
       'Offer Number', 'First Registration', 'Body Color', 'Paint Type',
       'Body Color Original', 'Upholstery', 'Body', 'Nr. of Doors',
       'Nr. of Seats', 'Model Code', 'Gearing Type', 'Displacement',
       'Cylinders', 'Weight', 'Drive chain', 'Fuel', 'Consumption',
       'CO2 Emission', 'Emission Class', '\nComfort & Convenience\n',
       '\nEntertainment & Media\n', '\nExtras\n', '\nSafety & Security\n',
       'description', 'Emission Label', 'Gears', 'Country version',
       'Electricity consumption', 'Last Service Date', 'Other Fuel Types',
       'Availability', 'Last Timing Belt Service Date', 'Available from'],
      dtype='object')

In [694]:
columns = df.columns
messy_df = pd.DataFrame(data = [], columns = columns, index = [0])
clean_df = clean_columns(messy_df)
clean_df.columns.tolist()

['url',
 'make_model',
 'short_description',
 'body_type',
 'price',
 'vat',
 'km',
 'registration',
 'prev_owner',
 'k_w',
 'hp',
 'type',
 'previous_owners',
 'next_inspection',
 'inspection_new',
 'warranty',
 'full_service',
 'non_smoking_vehicle',
 'null',
 'make',
 'model',
 'offer_number',
 'first_registration',
 'body_color',
 'paint_type',
 'body_color_original',
 'upholstery',
 'body',
 'nr_of_doors',
 'nr_of_seats',
 'model_code',
 'gearing_type',
 'displacement',
 'cylinders',
 'weight',
 'drive_chain',
 'fuel',
 'consumption',
 'co_2_emission',
 'emission_class',
 'comfort_&_convenience',
 'entertainment_&_media',
 'extras',
 'safety_&_security',
 'description',
 'emission_label',
 'gears',
 'country_version',
 'electricity_consumption',
 'last_service_date',
 'other_fuel_types',
 'availability',
 'last_timing_belt_service_date',
 'available_from']

In [695]:
df.columns = clean_df.columns.tolist()
df.columns

Index(['url', 'make_model', 'short_description', 'body_type', 'price', 'vat',
       'km', 'registration', 'prev_owner', 'k_w', 'hp', 'type',
       'previous_owners', 'next_inspection', 'inspection_new', 'warranty',
       'full_service', 'non_smoking_vehicle', 'null', 'make', 'model',
       'offer_number', 'first_registration', 'body_color', 'paint_type',
       'body_color_original', 'upholstery', 'body', 'nr_of_doors',
       'nr_of_seats', 'model_code', 'gearing_type', 'displacement',
       'cylinders', 'weight', 'drive_chain', 'fuel', 'consumption',
       'co_2_emission', 'emission_class', 'comfort_&_convenience',
       'entertainment_&_media', 'extras', 'safety_&_security', 'description',
       'emission_label', 'gears', 'country_version', 'electricity_consumption',
       'last_service_date', 'other_fuel_types', 'availability',
       'last_timing_belt_service_date', 'available_from'],
      dtype='object')

In [696]:
# clean_columns(df) # simpler code
# df=clean_columns(df)

## Checking the number of unique values in dataframe

In [697]:
# df.nunique()  #  The error "unhashable type: 'list'" is arrising because of  the values of list type.

In [698]:
# df.nunique(axis=1, dropna=True)  #  The error "unhashable type: 'list'" is arrising because of  the values of list type.

### Look at the counts of columns that have OBJECT datatype

In [699]:
df.select_dtypes('O').sort_index(axis=1).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 52 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   availability                   635 non-null    object
 1   available_from                 272 non-null    object
 2   body                           15859 non-null  object
 3   body_color                     15322 non-null  object
 4   body_color_original            12160 non-null  object
 5   body_type                      15859 non-null  object
 6   co_2_emission                  14111 non-null  object
 7   comfort_&_convenience          14999 non-null  object
 8   consumption                    14013 non-null  object
 9   country_version                7586 non-null   object
 10  cylinders                      10239 non-null  object
 11  description                    15919 non-null  object
 12  displacement                   15423 non-null  object
 13  d

### Assign the Columns (Features) of object data type as** **``"object_col"``

In [700]:
# df.columns[df.dtypes == 'O']
# df.select_dtypes('O').columns
# df.describe(include='O').columns

object_col = df.select_dtypes('O').sort_index(axis=1).columns
object_col

Index(['availability', 'available_from', 'body', 'body_color',
       'body_color_original', 'body_type', 'co_2_emission',
       'comfort_&_convenience', 'consumption', 'country_version', 'cylinders',
       'description', 'displacement', 'drive_chain', 'electricity_consumption',
       'emission_class', 'emission_label', 'entertainment_&_media', 'extras',
       'first_registration', 'fuel', 'full_service', 'gearing_type', 'gears',
       'hp', 'inspection_new', 'km', 'last_service_date',
       'last_timing_belt_service_date', 'make', 'make_model', 'model',
       'model_code', 'next_inspection', 'non_smoking_vehicle', 'nr_of_doors',
       'nr_of_seats', 'null', 'offer_number', 'other_fuel_types', 'paint_type',
       'prev_owner', 'previous_owners', 'registration', 'safety_&_security',
       'short_description', 'type', 'upholstery', 'url', 'vat', 'warranty',
       'weight'],
      dtype='object')

In [701]:
numeric_col = df.select_dtypes(exclude='O').sort_index(axis=1).columns
numeric_col

Index(['k_w', 'price'], dtype='object')

## Checking the descriptive values of df

### Checking the descriptive values of object columns

In [702]:
df.describe(include='O').T

count unique  \
url                            15919  15919   
make_model                     15919      9   
short_description              15873  10001   
body_type                      15859      9   
vat                            11406      2   
km                             15919   6690   
registration                   15919     48   
prev_owner                      9091      4   
hp                             15919     81   
type                           15917    169   
previous_owners                 9279    102   
next_inspection                 3535   1399   
inspection_new                  3932    202   
warranty                       10499    515   
full_service                    8215    122   
non_smoking_vehicle             7177     94   
null                           15919      1   
make                           15919      3   
model                          15919      9   
offer_number                   12744  11440   
first_registration             14322      4   
body_color                     15322     14   
paint_type                     10147      3   
body_color_original            12160   1927   
upholstery                     12199     46   
body                           15859      9   
nr_of_doors                    15707      6   
nr_of_seats                    14942      6   
model_code                      4978    232   
gearing_type                   15919      3   
displacement                   15423     77   
cylinders                      10239      7   
weight                          8945    434   
drive_chain                     9061      3   
fuel                           15919     77   
consumption                    14013    881   
co_2_emission                  14111    123   
emission_class                 12898      7   
comfort_&_convenience          14999   6198   
entertainment_&_media          14545    346   
extras                         12957    659   
safety_&_security              14937   4443   
description                    15919  12874   
emission_label                  3985      6   
gears                          11207     10   
country_version                 7586     23   
electricity_consumption          137      1   
last_service_date                566    266   
other_fuel_types                 880      1   
availability                     635     15   
last_timing_belt_service_date     16     15   
available_from                   272     46   

                                                                             top  \
url                            https://www.autoscout24.com//offers/audi-a1-sp...   
make_model                                                               Audi A3   
short_description                              SPB 1.6 TDI 116 CV S tronic Sport   
body_type                                                                 Sedans   
vat                                                               VAT deductible   
km                                                                         10 km   
registration                                                                 -/-   
prev_owner                                                      1 previous owner   
hp                                                                         85 kW   
type                                     [, Used, , Diesel (Particulate Filter)]   
previous_owners                                                            \n1\n   
next_inspection                                                      \n04/2022\n   
inspection_new                                             [\nYes\n, \nEuro 6\n]   
warranty                                                    [\n, \n, \nEuro 6\n]   
full_service                                             [\n, \n, \n4 (Green)\n]   
non_smoking_vehicle                                                     [\n, \n]   
null                                                                          []   
make                       

### Checking the descriptive values of numeric columns

In [703]:
df.describe(exclude="O").T

count      mean      std    min       25%       50%       75%  \
price 15919.000 18019.897 7386.169 13.000 12850.000 16900.000 21900.000   
k_w       0.000       NaN      NaN    NaN       NaN       NaN       NaN   

            max  
price 74600.000  
k_w         NaN

# Data Cleaning

## Defining a function to check null values, number of uniques and value counts for each column

In [704]:
def first_look(col):
    print('column name : ', col)
    print('--------------------------------')
    print('Per_of_Nulls   : ', '%', round(df[col].isnull().sum() / df.shape[0]*100, 2))
    print('Num_of_Nulls   : ', df[col].isnull().sum())
    print('Num_of_Uniques : ', df[col].nunique())
    print("Value_counts :\n",df[col].value_counts(dropna = False))
    print()
    print("Value_counts with sort index :\n",df[col].value_counts(dropna = False).sort_index())
    print("##"*20)
    print()

In [705]:
# Checking the informations for each column using defined function first_look

for col in df.columns :
    first_look(col)

column name :  url
--------------------------------
Per_of_Nulls   :  % 0.0
Num_of_Nulls   :  0
Num_of_Uniques :  15919
Value_counts :
 https://www.autoscout24.com//offers/audi-a1-sportback-1-4-tdi-s-tronic-xenon-navi-klima-diesel-black-bdab349a-caa5-41b0-98eb-c1345b84445e                  1
https://www.autoscout24.com//offers/opel-insignia-bus-inno-2-0-cdti-aut-navi-bi-xe-led-e6-diesel-black-8710b253-dd98-4ed0-9d46-cbad2d79532b                1
https://www.autoscout24.com//offers/opel-insignia-sport-tourer-opel-sport-tourer-1-6-cdti-136-ch-a-c-diesel-silver-10b7463b-9f56-4348-af35-df3432b2b585    1
https://www.autoscout24.com//offers/opel-insignia-2-0cdti-st-business-innovation-navi-bt-diesel-black-ba5a3368-dedc-4e51-9127-3976b26e93c2                 1
https://www.autoscout24.com//offers/opel-insignia-sport-tourer-opel-sport-tourer-1-6-cdti-136-ch-a-c-diesel-silver-7b52f054-78d3-4d5e-9370-8a7d081d7edd    1
                                                                               

TypeError: unhashable type: 'list'

## Changing data types of columns containing list type values to string

In [ ]:
# Changing data types of columns containing list type values to string

df = df.applymap(lambda x: ','.join(map(str, x)) if isinstance(x, list) else x)
df

## First cleaning of df

There were so many strange values that should be cleaned.

Firstly, "\n" values was cleaned.

In [ ]:
# Cleaning "\n" character and strip "," character from both side in df columns

df= df.apply(lambda x: x.str.replace("\n","").str.strip(",") if x.dtype == "O" else x)
df

### Checking columns with first_look function after first cleaning of df 

In [ ]:
# Checking the informations for each column using defined function first_look

for col in df.columns :
    first_look(col)

## First conclusion about df after checking the first_look results and descriptive values of each column 

After checking the first_look results and descriptive values of each column, it was decided

to replace empty rows with np.nan

to replace "[],[],[]" values with np.nan

to drop columns containing greater than or equal %95 null values

to apply data cleaning column by column.

In [ ]:
df.head()

### Detect strange values apart from the NaN Values.(isin(),Count(),Sum(),Any())

In [ ]:
# strip - Strip strange values apart from beginning and end
# df = df.applymap(lambda x: x.strip("\\n\n,][ '-.") if isinstance(x, str) else x, na_action='ignore')

### Replacing empty rows with np.nan

In [ ]:
# df["non_smoking_vehicle"][df.non_smoking_vehicle.str.contains(r'^\s*$')==True] = np.nan
# df["non_smoking_vehicle"].isnull().sum()

In [ ]:
# (df.non_smoking_vehicle.str.contains(r'^\s*$')).isnull().sum()

In [ ]:
# df.isin(['']).sum()

In [ ]:
# df[df.isin(['', '[],[],[]'])]= np.nan

In [ ]:
# df.isnull().sum()

In [ ]:
# df.non_smoking_vehicle.replace(to_replace=r'^\s*$', value=np.nan, regex=True).isnull().sum()

### Replacing rows containing only empt list with np.nan

In [ ]:
# df.other_fuel_types.value_counts(dropna=False)

In [ ]:
# df[df.isin(['','[]','','[],[]','[],[],[]'])]=np.nan
# df.isnull().sum()

In [ ]:
# df.isin(['[],[],[]']).sum()

In [ ]:
# df.other_fuel_types.replace(to_replace=r'\[\]', value=np.nan, regex=True).isnull().sum()

In [ ]:
# df

********************************

### Replacing empty rows with np.nan

In [ ]:
# Checking columns having empty rows

df.isin(['']).sum()

In [ ]:
# replacing empty rows with np.nan (1st method)

df[df.isin([''])] = np.nan

In [ ]:
# 2nd method to replace empty rows with np.nan

df.replace(to_replace=r'^\s*$', value=np.nan, regex=True).isnull().sum()

### Replacing rows containing only empt list with np.nan

In [ ]:
# Checking the columns having "[],[],[]" values

df.isin(['[],[],[]']).sum()

In [ ]:
# replacing rows containing "[],[],[]" with np.nan (1st method)

df[df.isin(['[],[],[]'])] = np.nan

In [ ]:
# 2nd method to replace "[],[],[]" values with np.nan

df.replace(to_replace=r'\[\]', value=np.nan, regex=True)

## Column by column data cleaning

### Drop columns containing greater than or equal %95 null values

In [ ]:
# %95 üzerinde null olan kolonlar

print(df.shape)
for col in df.columns :
    if (df[col].isnull().sum() / df.shape[0]*100 ) >= 95:
        print(col)
        print('Per_of_Nulls   : ', '%', round(df[col].isnull().sum() / df.shape[0]*100, 2))
        print("-"*25)

In [ ]:
#Alternative code
# df.isna().sum()
# df.isna().sum().div(len(df))*100
# (df.isna().sum().div(df.shape[0])*100).sort_values(ascending=False)

In [ ]:
# Dropping columns having greater than or equal to 95% null values
print(df.shape)
for col in df.columns :
    if (df[col].isnull().sum() / df.shape[0]*100 ) >= 95:
        df.drop(col, axis=1, inplace=True)
print(df.shape)

In [ ]:
# (df.isna().sum().div(df.shape[0]))*100  # alternatif null oranı bulma kodu

In [ ]:
# all values in the "kW" column are null. Because of that it will be dropped
# all values in the "null" column are an empty list. because of that it will be dropped

# df.drop(columns=["k_w","null"],inplace=True)

In [ ]:
len(df.columns)

In [ ]:
import re

In [ ]:
def get_diff_category_column(Series:pd.Series, exclude=''',/\n''', pattern=r'''[,\n]| /''', strip='''\n' "!?|.,*+-_/][\\n''', diff_value=list()):  
    diff_value.clear()
    column = Series.dropna().apply(str).str.strip(strip)
    for row in column:
        if not any(x in exclude for x in row) and row not in diff_value:
            diff_value.append(row)
        else:
            for data in map(lambda x: x.strip(strip), filter(None, re.split(pattern, row))):
                if data not in diff_value:
                    diff_value.append(data)
    return dict(enumerate(sorted(diff_value)))

In [ ]:
df.shape

In [ ]:
df.columns

## Data cleaning column by column  and fixing data types

Now, let's examine all columns as generally:

### url column

In [ ]:
first_look("url")

This column will be dropped as all rows are unique and the information in the row is not required.

In [ ]:
df.drop("url", axis=1,inplace=True)

In [ ]:
df.shape

### make_model column

In [ ]:
df.make_model

In [ ]:
first_look("make_model")

In [ ]:
get_diff_category_column(df.make_model)

In [ ]:
df[["make_model","make","model"]].sample(20)

In [ ]:
df[["make_model","make","model"]].isna().sum()

In [ ]:
df[["make_model","make","model"]].value_counts()

Since the "make" and "model" columns and "make_model" column contain the same values, let's drop the "make" and "model" columns.

In [ ]:
df.drop(columns=["make","model"],inplace=True)

In [ ]:
df.shape

### short_description column

In [ ]:
df.short_description

In [ ]:
first_look("short_description")

In [ ]:
df.groupby("make_model")["short_description"].value_counts(dropna=False).head(50)

> It can be deducted after looking at other columns. will be decided later.
> It can be used to fill other columns.

### body_type column

In [ ]:
df.body_type

In [ ]:
first_look("body_type")

In [ ]:
df[["body_type","body"]].isnull().sum()

In [ ]:
df[["body_type","body"]].value_counts(dropna=False)

In [ ]:
df.groupby("make_model")["body_type"].value_counts(dropna=False)

In [ ]:
df[df.body_type.isnull()].isnull().sum()

In [ ]:
# drop body column (same as body_type)

df.drop(columns="body", inplace=True)

In [ ]:
df.shape

### price column

In [ ]:
df.price

In [ ]:
first_look("price")

In [ ]:
df["price"]=df.price.astype(float)

In [ ]:
df.price

### vat column

In [ ]:
df.vat

In [ ]:
first_look("vat")

In [ ]:
get_diff_category_column(df.vat)

Since this column contains legal information and approximately 1/3 of it is null, it is more appropriate to drop it.

In [ ]:
df.drop("vat", axis=1, inplace=True)

In [ ]:
df.shape

### km column

In [ ]:
df.km

In [ ]:
first_look("km")

In [ ]:
#Cleaning "km" and "," characters in km column and fixing data type : 1st method
df["km"] = df.km.str.strip(" km").str.replace(",", "").replace(r"-", np.nan, regex=True).astype(float)

In [ ]:
# 2nd method
# df["km"] = df.km.str.replace("," , "").str.extract(r"(\d+)").astype(float)

In [ ]:
#checking the km column's values after cleaning 
df.km.value_counts(dropna=False)

### registration and first_registration columns

In [ ]:
df.registration

In [ ]:
first_look("registration")

In [ ]:
first_look("first_registration")

In [ ]:
df[["registration","first_registration"]].value_counts(dropna=False)

In [ ]:
df.groupby(["make_model","first_registration","registration"])["price"].apply(list)

In [ ]:
df.groupby(["make_model","first_registration","registration"])["price"].describe().tail(30)

In [ ]:
# Because both columns have the same values, let's drop the registration column:
df.drop(columns="registration",inplace=True)

In [ ]:
df.shape

### prev_owner and previous_owners columns

In [ ]:
# Checking prev_owner column
first_look("prev_owner")

In [ ]:
# Checking previous_owners column

first_look("previous_owners")

In [ ]:
df[["prev_owner","previous_owners", "consumption", "co_2_emission"]]

In [ ]:
df["prev_owner"] = df.prev_owner.str.extract("(\d)").astype(float)

In [ ]:
df.prev_owner.value_counts(dropna=False)

### hp column

In [ ]:
# Checking the hp column
first_look("hp")

In [ ]:
# Extracting hp values from hp column and assign them to hp_kw column and fixing data type

df["hp_kw"] = df.hp.str.extract(r"(\d+)").astype(float)

In [ ]:
first_look("hp_kw")

In [ ]:
df.hp_kw.head()

In [ ]:
# Dropping hp column
df.drop("hp", axis=1, inplace=True)

In [ ]:
df.hp_kw.isnull().sum()

In [ ]:
df.shape

### type column

In [ ]:
# Checking the type column

first_look("type")

In [ ]:
# Checking type and fuel column together

df[["type","fuel"]].head(50)

In [ ]:
# Extracting necessary values from type column : 1st method


df["type"] = df.type.str.extract(r"(.+),,")
df.type.value_counts()

In [ ]:
# Extracting necessary values from type column : 2nd method

# df.type.str.split(",").str[1].value_counts()

In [ ]:
# Checking null values after cleaning

df.type.isnull().sum()

There are only 2 null values in type column.

In [ ]:
# NOT: null değerlere sonra bakılacak.

### next_inspection column

In [ ]:
# Checking the next_inspection column

first_look("next_inspection")

In [ ]:
# Checking next_inspection and inspection_new columns together

df[["next_inspection","inspection_new"]].head(50)

There isn't any similarities between next_inspection and inspection_new columns. So these columns will be examined seperately.

In [ ]:
# Extracting date information from next_inspection columns and assign them to a new column named next_insp_date

df["next_insp_date"] = df.next_inspection.str.extract(r"(\d+/\d{0,4})")

In [ ]:
df["next_insp_date"].isnull().sum()

There are 12384 null values.

next_inspection column can be used to fill missing values in other columns. So it will drop later.

### inspection_new column

In [ ]:
# Checking the inspection_new column

first_look("inspection_new")

In [ ]:
# Extracting necessary information from inspection_new column and assign them into a new column
# named insp_new

df["insp_new"] = df.inspection_new.str.extract(r"(Yes)")

In [ ]:
df["insp_new"].isnull().sum()

There are 11987 null values.

inspection_new column can be used to fill missing values in other columns. So it will drop later.

### warranty column

In [ ]:
# Checking the warranty column

first_look("warranty")

In [ ]:
# Extracting necessary values from warranty column and assign them into a new column named warranty_month
# fixing data type

df["warranty_month"] = df.warranty.str.extract(r"(\d+)\s*month").astype(float)

In [ ]:
df["warranty_month"].isnull().sum()

warranty column can be used to fill other columns. Then can be dropped.

There are 11066 null values.

### full_service column

In [ ]:
# Checking the full_service column

first_look("full_service")

In [ ]:
# full_service column can be used to fill other columns' missing values. So it will drop later.

# df.drop(columns="full_service", inplace=True)

In [ ]:
df[df.full_service.notna()][["full_service","co_2_emission","emission_class", "emission_label"]].sample(50)

### non_smoking_vehicle column

In [ ]:
# Checking the non_smoking_vehicle column

first_look("non_smoking_vehicle")

This column includes some informations of the other columns.

### offer_number column

In [ ]:
# Checking the offer_number column

first_look("offer_number")

In [ ]:
df.offer_number.sample(50)

In [ ]:
# As offer_number column doesn't have any usefull/related data with car price,
# it was decided to drop this column.

df.drop(columns="offer_number",inplace=True)

In [ ]:
df.shape

### body_color

In [ ]:
# Checking the body_color column

first_look("body_color")

### paint_type column

In [ ]:
# Checking the paint_type column

first_look("paint_type")

### body_color_original column

In [ ]:
# Checking the body_color_original column

first_look("body_color_original")

In [ ]:
df[["body_color", "paint_type","body_color_original"]].sample(50)

In [ ]:
# The number of null values in body_color column

df.body_color.isnull().sum()

In [ ]:
# Checking how many of the null values in body_color column are not null in body_color_original column

df[df.body_color.isnull()]["body_color_original"].notna().sum()

As a result of the analysis of above, it may be possible to fill in some null values, but it is more useful to drop this column given that it is in a different language.

In [ ]:
df.drop("body_color_original", axis=1, inplace=True)

In [ ]:
df.shape

### upholstery column

In [ ]:
#Checking the upholstery column

first_look("upholstery")

In [ ]:
df.upholstery.str.extract(r"(.+),(.+)")

In [ ]:
# creating upholstery type list from upholstery column values using extract, dropna and unique methods.

type_list = list(df.upholstery.str.extract(r"(.+),(.+)")[0].dropna().unique())
type_list

In [ ]:
# creating upholstery color list from upholstery column values using extract, dropna and unique methods.

color_list = list(df.upholstery.str.extract(r"(.+),\s*(.+)")[1].dropna().unique())
color_list 

In [ ]:
# splitting upholstery column values into two column

df[["upholstery_type","upholstery_color"]] = df.upholstery.str.split(", ", expand=True)

In [ ]:
# As it is seen below output there are some color values in type column. So they need to be fixed.

df["upholstery_type"].value_counts(dropna=False)

In [ ]:
# Checking upholstery_color values whether there is upholstery_type value in it.

df["upholstery_color"].value_counts(dropna=False)

In [ ]:
# Determining color values in the upholstery_type column

df.upholstery_type[df.upholstery_type.isin(color_list)]

In [ ]:
# Assigning the color values which are in upholstery_type column to upholstery_color column

df.upholstery_color.fillna(df.upholstery_type[df.upholstery_type.isin(color_list)], inplace=True)

In [ ]:
# Checking the values in upholstery_color column

df["upholstery_color"].value_counts(dropna=False)

In [ ]:
# Replacing color values which are in upholstery_type column with null. 

df.upholstery_type[df.upholstery_type.isin(color_list)] = np.nan

In [ ]:
# Checking the values in upholstery_type column

df.upholstery_type.value_counts(dropna=False)

In [ ]:
df[["upholstery_type","upholstery_color"]].notna().sum()

In [ ]:
df[["upholstery_type","upholstery_color"]].isnull().sum()

In [ ]:
# dropping upholstery column

df.drop(columns="upholstery", inplace=True)

In [ ]:
df.shape

### nr_of_doors column

In [ ]:
# Checking the nr_of_doors column

first_look("nr_of_doors")

In [ ]:
# fixing data type

df["nr_of_doors"]= df.nr_of_doors.astype(float)

### nr_of_seats column

In [ ]:
# Checking the nr_of_seats column

first_look("nr_of_seats")

In [ ]:
# fixing data type

df["nr_of_seats"]= df.nr_of_seats.astype(float)

### model_code column

In [ ]:
# Checking the model_code column

first_look("model_code")

In [ ]:
# As there isn't any usefull/related information about car price,
# it was decided to drop model_code column.

df.drop(columns="model_code", inplace=True)

In [ ]:
df.shape

### gearing_type column

In [ ]:
# Checking the gearing_type column

first_look("gearing_type")

### displacement column

In [ ]:
# Checking the displacement column

first_look("displacement")

In [ ]:
# Replacing "," character with "" and extracting digits from displacement column.

df["displacement"] = df.displacement.str.replace(",", "").str.extract(r"(\d+)").astype(float)

In [ ]:
# Renaming the displacement column as displacement_cc

df.rename(columns={"displacement":"displacement_cc"}, inplace=True)

In [ ]:
df.displacement_cc.isnull().sum()

### cylinders columns

In [ ]:
# Checking the cylinders column

first_look("cylinders")

In [ ]:
# fixing data type

df["cylinders"] = df.cylinders.astype(float)

### weight column

In [ ]:
# Checking the weight column

first_look("weight")

In [ ]:
# Replacing "," character with "", extracting digits from weight column and assign as weight_kg column.

df["weight_kg"] = df.weight.str.replace(",", "").str.extract(r"(\d+)").astype(float)

In [ ]:
# drop weight column

df.drop(columns="weight", inplace=True)

In [ ]:
df.shape

In [ ]:
df.weight_kg.isnull().sum()

### drive_chain column

In [ ]:
# Checking the drive_chain column

first_look("drive_chain")

### fuel column

In [ ]:
# Checking the fuel column

first_look("fuel")

In [ ]:
# For further analysis it will be better to categorize fuel column.
# Defining lpg, diesel, bensine and electirc/hybrid lists with the related values in fuel column

lpg = ["lpg", "CNG", "Domestic gas H", "Biogas"]
diesel = ["Diesel", "Biodiesel"]
benzine =["Gasoline", "Benzine", "Regular", "Super", "Other"]
electric_hybride = ["Electric"]

In [ ]:
# Defining a function to categorize fuel column

def fuelcategory(row) :
    if any(x in row for x in lpg):
        return "LPG/CNG"
    elif any(x in row for x in electric_hybride) :
        return "Electric/Hybride"
    elif any(x in row for x in diesel) :
        if "Particulate Filter" in row :
            return "Diesel (Particulate Filter)"
        else :
            return "Diesel"
    else :
        if "Particulate Filter" in row :
            return "Benzine (Particulate Filter)"
        else :
            return "Benzine"
   

In [ ]:
# Application of defined function to fuel column

df["fuel"] = df.fuel.apply(fuelcategory)

In [ ]:
# Checking the value counts of fuel column after categorization process

df.fuel.value_counts()

### consumption column

In [ ]:
# Checking the consumption column

first_look("consumption")

In [ ]:
# Cleaning consumption column and taking fuel consumption values into separate columns
# Changing data types of new columns

df["consumption"]=df["consumption"].str.replace("[\[\]']?","")
df["consumption_comb"] = df.consumption.str.extract(r"(\d*\.\d+|\d+).+\(comb\)").astype(float)
df["consumption_city"] = df.consumption.str.extract(r"[.+]?,(\d*\.\d+|\d+).+\(city\)").astype(float)
df["consumption_country"] = df.consumption.str.extract(r".+,(\d*\.\d+|\d+).+\(country\)").astype(float)

In [ ]:
# Comparing new columns' values with consumptions' values

df[["consumption","consumption_comb","consumption_city","consumption_country"]].sample(50)

In [ ]:
# Checking the number of null values in new columns

df[["consumption_comb","consumption_city","consumption_country"]].isnull().sum()

In [ ]:
# Checking the number of non-null values in new columns

df[["consumption_comb","consumption_city","consumption_country"]].notna().sum()

In [ ]:
# dropping consumption column

df.drop(columns="consumption", inplace=True)

In [ ]:
df.shape

### co_2_emission column

In [ ]:
# Checking the co_2_emission column

first_look("co_2_emission")

In [ ]:
# Replacing "," character with "", extracting digits from co_2_emission column.

df["co_2_emission"] = df.co_2_emission.str.replace(",", "").str.extract(r"(\d+)").astype(float)

In [ ]:
df.co_2_emission.isnull().sum()

In [ ]:
df.co_2_emission

In [ ]:
df.co_2_emission.value_counts()

### emission_class column

In [ ]:
# Checking the emission_class column

first_look("emission_class")

### comfort_&_convenience column

In [ ]:
# Checking the comfort_&_convenience column

first_look("comfort_&_convenience")

In [ ]:
get_diff_category_column(df["comfort_&_convenience"])

In [ ]:
# Renaming the comfort_&_convenience column as comfort_convenience

df.rename(columns={"comfort_&_convenience":"comfort_convenience"}, inplace=True)

### entertainment_&_media column

In [ ]:
# Checking the entertainment_&_media column

first_look("entertainment_&_media")

In [ ]:
get_diff_category_column(df["entertainment_&_media"])

In [ ]:
# Renaming the entertainment_&_media column as entertainment_media

df.rename(columns={"entertainment_&_media":"entertainment_media"}, inplace=True)

### extras column 

In [ ]:
# Checking the extras column

first_look("extras")

In [ ]:
get_diff_category_column(df["extras"])

### safety_&_security column

In [ ]:
# Checking the safety_&_security column

first_look("safety_&_security")

In [ ]:
get_diff_category_column(df["safety_&_security"])

In [ ]:
# Renaming the safety_&_security column as safety_security

df.rename(columns={"safety_&_security":"safety_security"}, inplace=True)

### description column

In [ ]:
# Checking the description column

first_look("description")

In [ ]:
# drop description column

df.drop(columns="description", inplace=True)

In [ ]:
df.shape

### emission_label column

In [ ]:
# Checking the emission_label column

first_look("emission_label")

It will be compared with other related columns and then it may be dropped. 

### gears columns

In [ ]:
# Checking the gears column

first_look("gears")

In [ ]:
# Fixing data type

df["gears"] =  df.gears.astype(float)

### country_version column

In [ ]:
# Checking the country_version column

first_look("country_version")

In [ ]:
df.groupby("country_version").price.mean()

In [ ]:
# drop country_version column

df.drop(columns="country_version", inplace=True)

In [ ]:
df.shape

## Conclusion after cleaning process

*********

## Saving cleaned df as "scout_car_cleaned.csv"

In [ ]:
df.to_csv("scout_car_cleaned.csv", index=False)

In [ ]:
pd.read_csv("scout_car_cleaned.csv")